### DISCUSSION TITLE
[Immunization] `Record all` form functionality | Code or Content Changes

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2139] [Ideas]
# Context

- Record all tasks allows to user to record the administration dates for all the tasks in a group e.g. record all
  at-birth tasks at once from a single form.
- This form will have 3 Questionnaire widget types.
    - **Checkbox** - This will visualize the Immunizations that can be recorded at the moment.
    - **Radio** - This is a control question. It checks whether all the immunizations have the same administration
      date.
    - **Datepicker** - Collects the administration date data.

## Checkbox question functionality

- Prepopulate the Answer Options using the `DUE` or `OVERDUE` task information. Generate the answer options using the
  following options.
- Using Answer Expressions
    - The idea is to use the **answer expressions** SDC functionality to load up the `Tasks` as the answer options.
    - Only load the Tasks in the group with the Task status as `DUE` and `OVERDUE`.
- Writing custom Kotlin code to generate the answer options.
- The answer options should have the following characteristics.
  - The `Questionnaire.item.answerOption.valueCoding.id` as the `Task.id`.
  - The `Questionnaire.item.answerOption.valueCoding.display` as the `Task.description`.
  - The `Questionnaire.item.answerOption.valueCoding.code` as the `Task.id`.

## Radio question functionality

- This is a static question on the Questionnaire. It's used as an `enableWhen` control question.
- If yes is selected then a single date question is shown for all the immunizations on the Checkbox question above.
    - The assumption is all the immunizations were administered on the same day and the date given will be used on
      the `Observation.effectiveDateTime` attribute of all the Observations generated [here](https://github.com/opensrp/fhircore/discussions/2142).
- If no is selected then a date question for each of the selected answerOption is shown i.e. if the answerOptions have 3
  immunizations selected then show 3 date questions 1 for each selected answerOption.
    - Each of the dates is then used on the Observation generated for the selected answerOption.

## Datepicker question functionality

- **Single date questions**
    - This is a static question. It is shown when the Radio question `Yes` option is selected.
    - If this question is answered the date selected will be used on the administration Observations for all the
      selected Tasks on the CheckBox Question.
- **Multiple date questions**
    - These are date fields generated for each answer option on the CheckBox Question above.
    - These should be shown when the Radio question `No` option is selected and the answer option corresponding to the
      date question is checked.

## Pending Decisions and  R&D

1. Test how the answer expressions work on Questionnaires to evaluate suitability on this form
2. R&D on how to define `enableWhen` expressions dynamically while launching the Questionnaires.

## Implementation note

- The idea is to build this form dynamically using the [SDC](http://build.fhir.org/ig/HL7/sdc/) capabilities provided by the Data capture library on the FHIR Android SDK
    - In case the R&D shows that this is not possible we can pivot to writing Kotlin code to generate the desired form.
      This decision should be made after discussions with the rest of the tech team. 
    - The R&D documentation should be linked to this discussion item for future reference.

**Comment by pld** [comment]
For the "How the Record All form should work | Code or Content Changes" section, let's clarify that the task is to attempt to do with content only, if there's a blocker, report back to discuss, if we can't resolve we'll figure out a code solution, but not without sign-off.
